File: src/backstop/admission.py
from __future__ import annotations

import asyncio
import itertools
import threading
import time
from collections import deque
from dataclasses import dataclass

from .aimd import AIMDController
from .config import BackstopConfig, Priority


@dataclass
class _Ticket:
    priority: Priority
    sequence: int
    enqueued_at: float


class PriorityGate:
    def __init__(self, config: BackstopConfig, aimd: AIMDController) -> None:
        self._config = config
        self._aimd = aimd
        self._active = 0
        self._counter = itertools.count()
        self._queues: dict[Priority, deque[_Ticket]] = {
            Priority.CRITICAL: deque(),
            Priority.DEFAULT: deque(),
            Priority.BACKGROUND: deque(),
        }
        self._condition = threading.Condition()
        self._async_condition = asyncio.Condition()

    @property
    def active(self) -> int:
        with self._condition:
            return self._active

    @property
    def depth(self) -> int:
        with self._condition:
            return sum(len(queue) for queue in self._queues.values())

    def acquire(self, priority: Priority, timeout: float | None = None) -> float:
        ticket = _Ticket(priority, next(self._counter), time.monotonic())
        effective = timeout if timeout is not None else self._config.queue_timeout
        with self._condition:
            self._queues[priority].append(ticket)
            self._condition.notify_all()
            while not self._can_admit(ticket):
                if effective is None:
                    self._condition.wait()
                else:
                    if not self._condition.wait(timeout=effective):
                        raise TimeoutError("gate acquire timed out")
            self._queues[priority].popleft()
            self._active += 1
            return time.monotonic() - ticket.enqueued_at

    def release(self) -> None:
        with self._condition:
            self._active = max(0, self._active - 1)
            self._condition.notify_all()

    async def aacquire(self, priority: Priority, timeout: float | None = None) -> float:
        ticket = _Ticket(priority, next(self._counter), time.monotonic())
        effective = timeout if timeout is not None else self._config.queue_timeout
        async with self._async_condition:
            self._queues[priority].append(ticket)
            self._async_condition.notify_all()
            while not self._can_admit(ticket):
                try:
                    if effective is None:
                        await self._async_condition.wait()
                    else:
                        await asyncio.wait_for(self._async_condition.wait(), effective)
                except asyncio.TimeoutError:
                    raise TimeoutError("gate acquire timed out")
            self._queues[priority].popleft()
            self._active += 1
            return time.monotonic() - ticket.enqueued_at

    async def arelease(self) -> None:
        async with self._async_condition:
            self._active = max(0, self._active - 1)
            self._async_condition.notify_all()

    def _can_admit(self, ticket: _Ticket) -> bool:
        if self._active >= self._aimd.current_limit:
            return False
        chosen = self._choose_ticket()
        return chosen is ticket

    def _choose_ticket(self) -> _Ticket | None:
        now = time.monotonic()
        critical = self._queues[Priority.CRITICAL]
        if critical:
            return critical[0]

        aged = [
            queue[0]
            for queue in self._queues.values()
            if queue and now - queue[0].enqueued_at >= self._config.starvation_after_seconds
        ]
        if aged:
            return min(aged, key=lambda item: item.sequence)

        for priority in (Priority.DEFAULT, Priority.BACKGROUND):
            queue = self._queues[priority]
            if queue:
                return queue[0]
        return None

File: ts/backstop/src/admission.ts
import type { Priority } from "./types.js";

interface Ticket {
  priority: Priority;
  sequence: number;
  enqueuedAt: number;
}

/**
 * Additive-increase/multiplicative-decrease concurrency controller.
 * Mirrors the Python AIMDController.
 */
export class AIMDController {
  private limit: number;
  private lastAdjustment: number;
  private readonly minLimit: number;
  private readonly maxLimit: number;
  private readonly increase: number;
  private readonly decreaseFactor: number;
  private readonly intervalMs: number;

  constructor(opts: {
    initial: number;
    min: number;
    max: number;
    increase?: number;
    decreaseFactor: number;
    intervalMs: number;
  }) {
    this.limit = opts.initial;
    this.minLimit = opts.min;
    this.maxLimit = opts.max;
    this.increase = opts.increase ?? 1;
    this.decreaseFactor = opts.decreaseFactor;
    this.intervalMs = opts.intervalMs;
    this.lastAdjustment = -opts.intervalMs;
  }

  get currentLimit(): number {
    return this.limit;
  }

  recordSuccess(): boolean {
    const now = Date.now();
    if (now - this.lastAdjustment < this.intervalMs) return false;
    if (this.limit >= this.maxLimit) return false;
    this.limit = Math.min(this.maxLimit, this.limit + this.increase);
    this.lastAdjustment = now;
    return true;
  }

  recordPressure(): boolean {
    const now = Date.now();
    if (now - this.lastAdjustment < this.intervalMs) return false;
    const next = Math.max(this.minLimit, Math.floor(this.limit * this.decreaseFactor));
    if (next === this.limit) return false;
    this.limit = next;
    this.lastAdjustment = now;
    return true;
  }

  applyExternalDecrease(pressure = 1): boolean {
    const factor = Math.max(this.decreaseFactor, Math.min(1, pressure));
    const next = Math.min(this.maxLimit, Math.max(this.minLimit, Math.floor(this.limit * factor)));
    if (next === this.limit) return false;
    this.limit = next;
    this.lastAdjustment = Date.now();
    return true;
  }
}

/**
 * Priority admission gate with starvation prevention.
 * Mirrors the Python PriorityGate.
 */
export class PriorityGate {
  private active = 0;
  private seq = 0;
  private readonly queues: Record<Priority, Ticket[]> = {
    critical: [],
    high: [],
    default: [],
    low: [],
    bulk: [],
  };
  private readonly aimd: AIMDController;
  private readonly starvationMs: number;
  private readonly timeoutMs: number | null;

  constructor(aimd: AIMDController, opts: { starvationMs: number; timeoutMs: number | null }) {
    this.aimd = aimd;
    this.starvationMs = opts.starvationMs;
    this.timeoutMs = opts.timeoutMs;
  }

  get depth(): number {
    return Object.values(this.queues).reduce((n, q) => n + q.length, 0);
  }

  acquire(priority: Priority): Promise<number> {
    return new Promise((resolve, reject) => {
      const ticket: Ticket = { priority, sequence: this.seq++, enqueuedAt: Date.now() };
      this.queues[priority].push(ticket);
      let interval: ReturnType<typeof setInterval> | null = null;
      let done = false;

      const cleanup = () => {
        done = true;
        if (interval) { clearInterval(interval); interval = null; }
      };

      const tryAdmit = () => {
        if (done) return;
        if (this.canAdmit(ticket)) {
          this.queues[priority].shift();
          this.active++;
          cleanup();
          resolve(Date.now() - ticket.enqueuedAt);
          return;
        }
        if (this.timeoutMs != null) {
          const elapsed = Date.now() - ticket.enqueuedAt;
          if (elapsed >= this.timeoutMs) {
            const idx = this.queues[priority].indexOf(ticket);
            if (idx >= 0) this.queues[priority].splice(idx, 1);
            cleanup();
            reject(new Error("gate acquire timed out"));
          }
        }
      };

      tryAdmit();
      if (!this.canAdmit(ticket) && this.timeoutMs != null) {
        interval = setInterval(() => {
          if (done) { cleanup(); return; }
          tryAdmit();
        }, Math.max(10, this.timeoutMs / 10));
      }
    });
  }

  release(): void {
    this.active = Math.max(0, this.active - 1);
  }

  private canAdmit(ticket: Ticket): boolean {
    if (this.active >= this.aimd.currentLimit) return false;
    return this.chooseTicket() === ticket;
  }

  private chooseTicket(): Ticket | null {
    if (this.queues.critical.length) return this.queues.critical[0];
    const now = Date.now();
    for (const q of [this.queues.high, this.queues.default, this.queues.low, this.queues.bulk]) {
      if (q.length && now - q[0].enqueuedAt >= this.starvationMs) return q[0];
    }
    for (const q of [this.queues.default, this.queues.low, this.queues.bulk]) {
      if (q.length) return q[0];
    }
    return null;
  }
}
